From 2e2f83a284615ad8780c9a069e36d752ae2243bb Mon Sep 17 00:00:00 2001
From: Harish Kasiviswanathan <Harish.Kasiviswanathan@amd.com>
Date: Fri, 2 Sep 2016 16:03:44 -0400
Subject: [PATCH 1871/2094] drm/amdgpu: Signal eviction fence

When amdgpu_amdkfd_fence.enable_signaling() is called, it means that GPU
scheduler has a job ready to evict a KFD BO. Notify KFD driver to
prepare for eviction.

Change-Id: I0a00a5df5e79f39e42d790842efacb67cc3e918b
Signed-off-by: Harish Kasiviswanathan <Harish.Kasiviswanathan@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd_fence.c | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd_fence.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd_fence.c
index 6fdd24c..7026781 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd_fence.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd_fence.c
@@ -117,13 +117,19 @@ static const char *amd_kfd_fence_get_timeline_name(struct fence *f)
  */
 static bool amd_kfd_fence_enable_signaling(struct fence *f)
 {
+	struct amdgpu_amdkfd_fence *fence = to_amdgpu_amdkfd_fence(f);
+
+	if (!fence)
+		return false;
+
 	if (fence_is_signaled(f))
 		return true;
 
-	/* TODO: If the fence is not signaled, call into KFD to schedule
-	 * work item that will prepare for KFD BO evictions
-	 */
-	return true;
+	if (!kgd2kfd->schedule_evict_and_restore_process(
+				(struct mm_struct *)fence->mm, f))
+		return true;
+
+	return false;
 }
 
 static int amd_kfd_fence_signal(struct fence *f)
-- 
2.7.4

